src/directdb.rs

```diff
@@ -0,0 +1,146 @@
+    record_type: String,
+}
+
+pub async fn get_objects_from_shard(
+    shard: u32,
+    conf: Config,
+    log: Logger,
+    obj_tx: crossbeam::Sender<SharkspotterMessage>,
+) -> Result<(), Error> {
+    let shard_host_name =
+        format!("{}.rebalancer-postgres.{}", shard, conf.domain);
+
+    debug!(log, "Connecting to {}", shard_host_name);
+    let (client, connection) = tokio_postgres::Config::new()
+        .host(shard_host_name.as_str())
+        .user("postgres")
+        .dbname("moray")
+        .keepalives_idle(std::time::Duration::from_secs(30))
+        .connect(NoTls)
+        .await
+        .map_err(|e| Error::new(ErrorKind::Other, e))?;
+
+    tokio::spawn(async move {
+        connection
+            .await
+            .map_err(|e| Error::new(ErrorKind::Other, e))?;
+        Ok::<(), Error>(())
+    });
+
+    let rows = client
+        .query_raw("SELECT * from manta where type='object'", vec![])
+        .await
+        .map_err(|e| Error::new(ErrorKind::Other, e))?;
+
+    pin_mut!(rows);
+    while let Some(row) = rows
+        .try_next()
+        .await
+        .map_err(|e| Error::new(ErrorKind::Other, e))?
+    {
+        let val_str: &str = row.get("_value");
+        let value: Value = serde_json::from_str(val_str)
+            .map_err(|e| Error::new(ErrorKind::Other, e))?;
+        let obj_id = object_id_from_manta_obj(&value)
+            .map_err(|e| Error::new(ErrorKind::Other, e))?;
+        let sharks = get_sharks_from_manta_obj(&value, &log)?;
+
+        trace!(log, "sharkspotter checking {}", obj_id);
+        if let Err(e) = sharks
+            .iter()
+            .filter(|s| conf.sharks.contains(&s.manta_storage_id))
+            .try_for_each(|s| {
+                trace!(log, "Found matching record: {:#?}", &row);
+                let moray_object: MorayMantaBucketObject =
+                    serde_postgres::from_row(&row).map_err(|e| {
+                        error!(
+                            log,
+                            "Error deserializing record as manta object: {}", e
+                        );
+                        Error::new(ErrorKind::Other, e)
+                    })?;
+
+                let etag = moray_object._etag.clone();
+                let manta_value_str = moray_object._value.as_str();
+                let manta_value: Value = serde_json::from_str(manta_value_str)
+                    .map_err(|e| Error::new(ErrorKind::Other, e))?;
+
+                debug!(log, "Sending value: {:#?}", manta_value);
+
+                let msg = SharkspotterMessage {
+                    manta_value,
+                    etag,
+                    shark: s.manta_storage_id.clone(),
+                    shard,
+                };
+
+                if let Err(e) = obj_tx.send(msg) {
+                    warn!(log, "Tx channel disconnected: {}", e);
+                    return Err(Error::new(ErrorKind::BrokenPipe, e));
+                }
+                Ok(())
+            })
```

Comment:
This is mostly just me trying to understand what's going on rather then critique.

My understanding:
This massive loop is iterating through each row in the results from the query, filtering by a specific storage_id (presumably the one of the shark that we want to evacuate in the context of the rebalancer at least), then it's extracting the necessary information from it in order to populate the `SharkspotterMessage` object, which it then forwards on to the client?

If my understanding is wrong, then disregard the rest.  My opinion, and it's optional, would be to break a good chunk of this loop out in to a function that we can also test by itself by faking up some data, and running it through the function and validating that we get what we expect at the other end.  I'd probably advocate that more strongly if this was going to be around indefinitely, but it's not, so I leave it up to you.

Reply:
You're understanding is correct.  I refactored some of this.